### ISSUE TITLE
ComponentRenderer.updateComponent should not recreate Component if itemConsumer is set

### ISSUE THREAD

**Issue by mperktold** [vaadin/flow#8637]
One constructor of `ComponentRenderer` takes a `SerializableSupplier<COMPONENT> componentSupplier` for creating a new, "empty" component, and a `SerializableBiConsumer<COMPONENT, SOURCE> itemConsumer` for updating a component according to the item data.

For many `ComponentRenderer`s, this is a very natural way of creating them.
For example:

```java
ComponentRenderer<Checkbox, Boolean> checkboxRenderer = new ComponentRenderer<>(Checkbox::new, Checkbox::setValue);
ComponentRenderer<Span, String> spanRenderer = new ComponentRenderer<>(Span::new, Span::setText);
```

In terms of behavior, I would expect these to be a shorthand for the following, much more verbose constuction:

```java
ComponentRenderer<Checkbox, Boolean> checkboxRenderer = new ComponentRenderer(
  value -> new Checkbox(value),
  (component, value) -> {
    ((Checkbox)component).setValue(value);
    return component;
  }
);
```

But instead, it behaves like:

```java
ComponentRenderer<Checkbox, Boolean> checkBoxRenderer = new ComponentRenderer<>(value -> new Checkbox(value));
ComponentRenderer<Span, String> spanRenderer = new ComponentRenderer<>(value -> new Span(value));
```

That is, when update component `ComponentRenderer.updateComponent(COMPONENT, SOURCE)` is called, the renderer recreates the component from scratch instead of updating the previously rendered instance.

I find this counter-intuitive because if I really wanted this behavior, I could as well use the constructor that takes a single `SerializableFunction<T, COMPONENT>` as shown above.
Instead, when providing an `itemConsumer`, I'm actually saying "this is how you should update the rendered component".

With this proposal implemented, each constructor would offer another behavior in terms of recreating vs. updating existing components (except a single function vs. a single supplier):

```java
// Create once, reuse and update later
ComponentRenderer(
    SerializableSupplier<COMPONENT> componentSupplier,
    SerializableBiConsumer<COMPONENT, SOURCE> itemConsumer)

// Always recreate
ComponentRenderer(
    SerializableSupplier<COMPONENT> componentSupplier)
ComponentRenderer(
    SerializableFunction<SOURCE, COMPONENT> componentFunction)

// Let componentUpdateFunction decide whether to recreate
ComponentRenderer(
    SerializableFunction<SOURCE, COMPONENT> componentFunction,
    SerializableBiFunction<Component, SOURCE, Component> componentUpdateFunction)
```

Currently, all constructors except the one receiveing a `componentUpdateFunction` always recreate the component from scratch.
All the others merely accept different function types for achieving the same behavior, so the only benefit you get in having all of them is that you can more likely use method references.

This is understandable from a historical point of view, because `updateComponent`, `componentUpdateFunction` and the corresponding constructor were added later in #4702. Before that, it wasn't possible to reuse the old component, so there was no other behavior to offer to begin with.

But now with the added flexibility, I feel that the behavior of the other constructors should be reconsidered to maximize the value we get from them.

I can only see the following potential problems with the proposed change:

### Backward compatibility

Changing the behavior might brake code that assumes the current behavior.
For example, the `itemConsumer` conditionally sets several component properties, it would now have to clear the others:

```java
ComponentRenderer<Span, String> spanRenderer = new ComponetRenderer<>(
  Span::new,
  (span, value) -> {
    span.setText(value);
    // show in bold if enclosed in '*'
    if (value.matches("\\*.*\\*")) {
      span.getStyle().set("font-weight", "bold");
    } else {
      // This would be required now but wasn't before
      span.getStyle().remove("font-weight");
    }
  }
);
```

It is hard to say how much code assumes the current behavior. It mostly depends on whether the behavior is explicitly guaranteed by JavaDoc of the corresponding constructor. It contains the following paragraph on the matter:

```
* Some components may support several rendered components at once, so
* different component instances should be created for each different item
* for those components.
```

It's unclear to me whether this is to say that the constructor will create a new component all the time, or to say that the provided functions should create different components for different items.

So to summarize, it might be OK to introduce a breaking change for this, or it might not.
If not, maybe the proposed behavior can be provided by a static factory method with a descriptive name, e.g.:

```java
public static <COMPONENT extends Component, SOURCE>
ComponentRenderer<COMPONENT, SOURCE> reusingRendered(
  SerializableSupplier<COMPONENT> componentSupplier,
  SerializableBiConsumer<COMPONENT, SOURCE> itemConsumer
) {
  ...
}
```

### Typing

Currently, the signature of `updateComponent` uses `Component` instead of `COMPONENT`, so it cannot use `itemConsumer` directly without casting the existing component first.

But #8636 proposes to change the signature, so this problem might be fixed.

**Comment by Legioth** [comment]
It seems like this analysis misses the point that reuse can only happen in the case of explicitly doing `reset(item)` or `resetAll()` on the data provider.

Since components are stateful, it's necessary to have one component instance for each active item. It's not practical to reuse component instances after an item is no longer active (i.e. scrolled out of view) because the logic that updates the rendering cannot know which of the previously rendered rows will be reused for a given new position. I've done a prototype that would enable doing so (see https://github.com/vaadin/vaadin-grid-flow/issues/765#issuecomment-643506064), but doing it in that way is a quite significant departure for how `ComponentRenderer` typically works.

**Comment by mperktold** [comment]
Right, the benefit would be for `refreshItem(T)` and `refreshAll()`.
However, those methods must be called wenever you change the underlying data source, e.g. by adding or removing items, so it could be worthwile. As you explained, inactive components shouldn't be reused, that makes sense to me.

Anyway, I don't meen to change the general strategy here.
`ComponentRenderer` already provides the method for updating an existing component, possibly reusing that component instance, and `ComponentDataGenerator.refreshData(T)` delegates to this method.

So basically, Vaadin is already able to reuse components for refreshing items, and that's great.

I'm only suggesting to use that strategy for one more constructor of `ComponentRenderer`.
More specifically, I suggest that passing an `itemConsumer` would behave like passing a `componentUpdateFunction` that changes the existing component according to the item data and always returns the same instance.

It's still up to the framework and to the specific components whether to call `createComponent` or `updateComponent`.

> I've done a prototype that would enable doing so (see vaadin/vaadin-grid-flow#765 (comment)), but doing it in that way is a quite significant departure for how ComponentRenderer typically works.

That's interesting, thanks for the pointer!
